### PR TITLE
Porting tgstation's seed extractor storage menu

### PR DESCRIPTION
**Pull request by 9600bauds**
Porting over https://github.com/tgstation/-tg-station/pull/2973 plus a little tweaking to allow for vending of more than one seed packet at a time.
Full credits to Kelenius for the code. Original PR submitted Mar 4, 2014, so hopefully there won't be any licensing issues.

ss 2015-03-23 at 02 59 32

### CODE REVIEW

File: code/game/machinery/seed_extractor.dm [hunk 1]
```diff
@@ -28,6 +29,29 @@
 
 obj/machinery/seed_extractor/attackby(var/obj/item/O as obj, var/mob/user as mob)
 
+	// Emptying a plant bag
+	if (istype(O,/obj/item/weapon/storage/bag/plants))
+		var/obj/item/weapon/storage/P = O
+		var/loaded = 0
+		for(var/obj/item/seeds/G in P.contents)
+			if(contents.len >= 999)
```
Comment (d3athrow) [inline code/game/machinery/seed_extractor.dm]: what an arbitrary amount

File: code/game/machinery/seed_extractor.dm [hunk 2]
```diff
@@ -60,7 +84,7 @@ obj/machinery/seed_extractor/attackby(var/obj/item/O as obj, var/mob/user as mob
 		S.use(1)
 		new /obj/item/seeds/grassseed(loc)
 
-	if(O)
+	if(O) //not sure what this portion of code is supposed to do - needs some commenting
```
Comment (d3athrow) [inline code/game/machinery/seed_extractor.dm]: If above O was deleted or used up on https://github.com/d3athrow/vgstation13/pull/3547/files#diff-211ff530d84779ca3328122cc4029712L60 it wouldnt want to try and continue to process.

File: code/game/machinery/seed_extractor.dm (hunk 1 above)
Comment (9600bauds) [inline code/game/machinery/seed_extractor.dm]: apparently some BYOND stuff breaks if it holds 1000 or more items, smartfridge.dm explains in more detail:
var/global/max_n_of_items = 999 // Sorry but the BYOND infinite loop detector doesn't look things over 1000.

File: code/game/machinery/seed_extractor.dm (hunk 2 above)
Comment (9600bauds) [inline code/game/machinery/seed_extractor.dm]: Kinda meant the rest of that process, I can't tell what it does at a glance or at a frowning stare

File: code/game/machinery/seed_extractor.dm (hunk 1 above)
Comment (ComicIronic) [inline code/game/machinery/seed_extractor.dm]: Then use max_n_of_items at this point in the code.

Comment (9600bauds) [inline code/game/machinery/seed_extractor.dm]: Moved the global to a define in setup.dm, for whatever god-forsaken reason it was being declared a second time in microwave.dm, which shouldn't even have been using it.

**Comment by dylanstrategie** [comment]
This looks really nice, and it's a straight port so it likely works, but just for information how does it handle seeds of the same type but with different stats ?

File: code/game/machinery/seed_extractor.dm [hunk 3]
```diff
@@ -74,3 +106,102 @@ obj/machinery/seed_extractor/attackby(var/obj/item/O as obj, var/mob/user as mob
...
+		var/obj/item/seeds/O = T
+		if (O.seed.display_name == href_list["name"] && O.seed.lifespan == href_list["li"] && O.seed.endurance == href_list["en"] && O.seed.maturation == href_list["ma"] && O.seed.production == href_list["pr"] && O.seed.yield == href_list["yi"] && O.seed.potency == href_list["pot"]) //Boy this sure is a long line, let's have this comment stretch it out even more!
+			O.loc = src.loc
+			amt--
+			if (amt <= 0) break
+
+	src.updateUsrDialog()
+	return
+
+obj/machinery/seed_extractor/proc/add(var/obj/item/seeds/O as obj)
```
Comment (ComicIronic) [inline code/game/machinery/seed_extractor.dm]: I suggest you rename this proc to something more specific, just for code sanity.

File: code/game/machinery/seed_extractor.dm [hunk 4]
```diff
@@ -74,3 +106,102 @@ obj/machinery/seed_extractor/attackby(var/obj/item/O as obj, var/mob/user as mob
...
+	if(contents.len >= MAX_N_OF_ITEMS)
+		usr << "<span class='notice'>\The [src] is full.</span>"
```
Comment (ComicIronic) [inline code/game/machinery/seed_extractor.dm]: usr is bad. You should pass a user arg to this proc, or make it a visible_message, or handle the code elsewhere.

File: code/game/machinery/seed_extractor.dm [hunk 5]
```diff
@@ -74,3 +106,102 @@ obj/machinery/seed_extractor/attackby(var/obj/item/O as obj, var/mob/user as mob
...
+				return
+			N.amount = max(N.amount - amt, 0)
+			if (N.amount <= 0)
+				piles -= N
+				del(N)
+			break
+
+	for (var/obj/T in contents)//Now we find the seed we need to vend
+		var/obj/item/seeds/O = T
+		if (O.seed.display_name == href_list["name"] && O.seed.lifespan == href_list["li"] && O.seed.endurance == href_list["en"] && O.seed.maturation == href_list["ma"] && O.seed.production == href_list["pr"] && O.seed.yield == href_list["yi"] && O.seed.potency == href_list["pot"]) //Boy this sure is a long line, let's have this comment stretch it out even more!
```
Comment (ComicIronic) [inline code/game/machinery/seed_extractor.dm]: This is a check you do more than twice, you might want to put it into a proc.

File: code/game/machinery/seed_extractor.dm [hunk 6]
```diff
@@ -74,3 +106,102 @@ obj/machinery/seed_extractor/attackby(var/obj/item/O as obj, var/mob/user as mob
...
+	src.lifespan = life
+	src.endurance = endur
+	src.maturation = matur
+	src.production = prod
+	src.yield = yie
+	src.potency = poten
+	src.amount = am
+
+/obj/machinery/seed_extractor/attack_hand(mob/user as mob)
+	user.set_machine(src)
```
Comment (ComicIronic) [inline code/game/machinery/seed_extractor.dm]: This should really be in interact.

File: code/game/machinery/seed_extractor.dm (hunk 3 above)
Comment (9600bauds) [inline code/game/machinery/seed_extractor.dm]: Can't hurt, you're right.

File: code/game/machinery/seed_extractor.dm (hunk 5 above)
Comment (9600bauds) [inline code/game/machinery/seed_extractor.dm]: I tried to make those Long Lines of Thorough If Checks into a proc, but they're trickier than it looks. First one compares href_list to a seed_pile datyum, second compares href_list to a seed/seed datum, and third compares the seed/seed datum to the seed_pile datum. Only way I could even get the proc to compile was passing each variable individually as a proc (plus a bonus var to see if it checked for name or display_name for checks 2 and 3) and have the proc just check if everything was equal, at which point, why even have a proc if it was just an if statement?

File: code/game/machinery/seed_extractor.dm (hunk 6 above)
Comment (9600bauds) [inline code/game/machinery/seed_extractor.dm]: I couldn't really understand what this did even after checking with the IRC, I just moved it to interact like you said. If it's still wrong please do tell.

**Comment by 9600bauds** [comment]
@Dylan: It displays seeds with different stats in a new line. This is extremely useful for mutating or splicing, it essentially lets you have an ark of noah-type quick seed lookup and pick the optimal ones.

@Comic: Thank you for the corrections, I'm actually glad to go through quality checking!

**Pull request merged**